Start with iterating through the text and seeing how many lines we have in our dataset 

In [1]:
import pandas as pd
import geonamescache
import numpy as np
import matplotlib.pyplot as plt
import re

In [2]:
count = 0
file = open("./data/headlines.txt", 'r')
for line in file.readlines():
    count += 1
    print(line)
file.close()
print(f'Total Headlines {count}')

Zika Outbreak Hits Miami

Could Zika Reach New York City?

First Case of Zika in Miami Beach

Mystery Virus Spreads in Recife, Brazil

Dallas man comes down with case of Zika

Trinidad confirms first Zika case

Zika Concerns are Spreading in Houston

Geneve Scientists Battle to Find Cure

The CDC in Atlanta is Growing Worried

Zika Infested Monkeys in Sao Paulo

Brownsville teen contracts Zika virus

Mosquito control efforts in St. Louis take new tactics with Zika threat

San Juan reports 1st U.S. Zika-related death amid outbreak

Flu outbreak in Galveston, Texas

Zika alert – Manila now threatened

Zika afflicts 7 in Iloilo City

New Los Angeles Hairstyle goes Viral

Louisiana Zika cases up to 26

Orlando volunteers aid Zika research

Zika infects pregnant woman in Cebu

Chicago's First Zika Case Confirmed

Tampa Bay Area Zika Case Count Climbs

Bad Water Leads to Sickness in Flint, Michigan

Baltimore plans for Zika virus

London Health Unit Tracks Mad Cow Disease

Zika cases in Viet

In [3]:
tester = 'New medicine wipes out Tuberculosis in Taipei' 

In [4]:
gc = geonamescache.GeonamesCache()
countries = gc.get_countries_by_names()

In [5]:
for country in countries.keys():
    if country in tester:
        print(country)

In [6]:
def check_countries(headline):
    countries_in_headline = []
    # Get List of Cities
    gc = geonamescache.GeonamesCache()
    cities = gc.get_countries_by_names()
    # Iterate Over List of Countries
    for country in countries.keys():
        if country in headline:
            countries_in_headline.append(country)
    if not countries_in_headline:
        return np.nan
    return countries_in_headline

In [7]:
def check_city(headline):
    cities_in_headline = []
    # Get List of Cities
    gc = geonamescache.GeonamesCache()
    cities = gc.get_cities()
    # Iterate Over List of Cities
    for city_index in cities.keys():
        if cities[city_index]['name'] in headline:
            cities_in_headline.append(cities[city_index]['name'])
    if not cities_in_headline:
        return np.nan
    return cities_in_headline

In [8]:
%%time
headline = []
country_hl = []
city_hl = []
file = open("./data/headlines.txt", 'r')
for counter, line in enumerate(file.readlines()):
    if counter % 25 == 0:
        print(f'{counter} completed out of 650')
    # remove 's
    line = line.replace("'s", '')
    # remove punctuation
    line = re.sub(r'[^\w\s]','',line)
    # remove new line
    line = line.replace('\n', '')
    headline.append(line)
    country_hl.append(check_countries(line))
    city_hl.append(check_city(line))
file.close()

0 completed out of 650
25 completed out of 650
50 completed out of 650
75 completed out of 650
100 completed out of 650
125 completed out of 650
150 completed out of 650
175 completed out of 650
200 completed out of 650
225 completed out of 650
250 completed out of 650
275 completed out of 650
300 completed out of 650
325 completed out of 650
350 completed out of 650
375 completed out of 650
400 completed out of 650
425 completed out of 650
450 completed out of 650
475 completed out of 650
500 completed out of 650
525 completed out of 650
550 completed out of 650
575 completed out of 650
600 completed out of 650
625 completed out of 650
CPU times: user 42.5 s, sys: 942 ms, total: 43.5 s
Wall time: 43.5 s


In [9]:
len(headline), len(country_hl), len(city_hl)

(650, 650, 650)

In [10]:
df = pd.DataFrame({'Headline': headline,
                  'Country': country_hl,
                  'City': city_hl})

In [11]:
df.head(25)

Headline   Country  \
0                            Zika Outbreak Hits Miami       NaN   
1                      Could Zika Reach New York City       NaN   
2                   First Case of Zika in Miami Beach       NaN   
3              Mystery Virus Spreads in Recife Brazil  [Brazil]   
4             Dallas man comes down with case of Zika       NaN   
5                   Trinidad confirms first Zika case       NaN   
6              Zika Concerns are Spreading in Houston       NaN   
7               Geneve Scientists Battle to Find Cure       NaN   
8               The CDC in Atlanta is Growing Worried       NaN   
9                  Zika Infested Monkeys in Sao Paulo       NaN   
10              Brownsville teen contracts Zika virus       NaN   
11  Mosquito control efforts in St Louis take new ...       NaN   
12  San Juan reports 1st US Zikarelated death amid...       NaN   
13                    Flu outbreak in Galveston Texas       NaN   
14                  Zika alert  Manila now threatened       NaN   
15                     Zika afflicts 7 in Iloilo City       NaN   
16               New Los Angeles Hairstyle goes Viral       NaN   
17                      Louisiana Zika cases up to 26       NaN   
18               Orlando volunteers aid Zika research       NaN   
19                Zika infects pregnant woman in Cebu       NaN   
20                  Chicago First Zika Case Confirmed       NaN   
21              Tampa Bay Area Zika Case Count Climbs       NaN   
22      Bad Water Leads to Sickness in Flint Michigan       NaN   
23                     Baltimore plans for Zika virus       NaN   
24          London Health Unit Tracks Mad Cow Disease       NaN   

                                                 City  
0                                             [Miami]  
1                         [York, York, New York City]  
2                                [Miami, Miami Beach]  
3                                       [Recife, Bra]  
4                                    [Dallas, Dallas]  
5                      [Trinidad, Trinidad, Trinidad]  
6                                       [Ho, Houston]  
7                                                 NaN  
8                                           [Atlanta]  
9                                          [Pau, Mon]  
10            [Brownsville, Brownsville, Brownsville]  
11                                                NaN  
12  [San Juan, San Juan, San, San Juan, San Juan, ...  
13                                        [Galveston]  
14                                      [Man, Manila]  
15                                      [Ilo, Iloilo]  
16                                      [Los Angeles]  
17                                                NaN  
18                                          [Orlando]  
19                                                NaN  
20                                          [Chicago]  
21                                       [Bay, Tampa]  
22                                 [Flint, Wa, Flint]  
23                                        [Baltimore]  
24                               [London, London, Un]

Check how many instances I have in the Country Column 


In [12]:
country_dict = {}
for x in df['Country']:
    # count null values
    if type(x) == float:
        country_dict['nan'] = country_dict.get('nan', 0) + 1 
    else:
        len_of_list = len(x)
        country_dict[str(len_of_list)] = country_dict.get(str(len_of_list), 0) + 1 

In [13]:
country_dict

{'nan': 634, '1': 16}

I have 16 headlines with a list of 1 country and 634 headlines with no countries 

In [14]:
city_dict = {}
my_list = []
for x in df['City']:
    # count null values
    if type(x) == float:
        city_dict['nan'] = city_dict.get('nan', 0) + 1 
        my_list.append('nan')
    else:
        len_of_list = len(x)
        city_dict[str(len_of_list)] = city_dict.get(str(len_of_list), 0) + 1 
        my_list.append(len_of_list)

In [15]:
city_dict

{'1': 260,
 '3': 81,
 '2': 172,
 'nan': 41,
 '6': 14,
 '7': 10,
 '5': 26,
 '8': 1,
 '4': 43,
 '9': 2}

In [16]:
df['city_len'] = my_list

In [17]:
df.loc[df['city_len'] == 9]

Headline Country  \
138            Spike of Pneumonia Cases in Springfield     NaN   
255  Lower Hospitalization in Richmond after Mumps ...     NaN   

                                                  City city_len  
138  [Springfield, Springfield, Springfield, Spring...        9  
255  [Richmond, Richmond, Man, Ho, Richmond, Richmo...        9

In [18]:
df.loc[138, 'City']

['Springfield',
 'Springfield',
 'Springfield',
 'Springfield',
 'Springfield',
 'Spring',
 'Springfield',
 'Springfield',
 'Springfield']

In [19]:
df.loc[255, 'City']

['Richmond',
 'Richmond',
 'Man',
 'Ho',
 'Richmond',
 'Richmond',
 'Richmond',
 'Richmond',
 'Richmond']

In [20]:
df.loc[255, 'Headline']

'Lower Hospitalization in Richmond after Mumps Vaccine becomes Mandatory'

It looks like a couple of things are happening here
1. There are multiple instances of Springfield and without having the state I am not sure which Springfield it is.  Same thing with Richmond
2. There are instances where I am grabbing the beginning of the word `Ho` from `Hospitalization` 

I am going to remove any instance where a city appears multiple times because if I make a guess which Springfield or Richmond it is, this will give us bad estimates. 

In [21]:
df.head()

Headline   Country  \
0                 Zika Outbreak Hits Miami       NaN   
1           Could Zika Reach New York City       NaN   
2        First Case of Zika in Miami Beach       NaN   
3   Mystery Virus Spreads in Recife Brazil  [Brazil]   
4  Dallas man comes down with case of Zika       NaN   

                          City city_len  
0                      [Miami]        1  
1  [York, York, New York City]        3  
2         [Miami, Miami Beach]        2  
3                [Recife, Bra]        2  
4             [Dallas, Dallas]        2

In [22]:
df.columns

Index(['Headline', 'Country', 'City', 'city_len'], dtype='object')

In [23]:
# Remove instances where the same city appears multiple times 
for index, row in df.iterrows():
    if type(row['city_len']) == int:
        if row['city_len'] > 1:
            city_count = []
            for city in row['City']:
                if city not in city_count:
                    city_count.append(city)
                else:
                    df.loc[index, 'City'] = np.nan

Veify that it worked.  Expecting the `City` column to be `NaN` for both rows 

In [24]:
df.loc[138, :]

Headline    Spike of Pneumonia Cases in Springfield
Country                                         NaN
City                                            NaN
city_len                                          9
Name: 138, dtype: object

In [25]:
df.loc[255, :]

Headline    Lower Hospitalization in Richmond after Mumps ...
Country                                                   NaN
City                                                      NaN
city_len                                                    9
Name: 255, dtype: object

Looks like my for loop worked, lets check how many rows this impacted 

In [26]:
city_dict = {}
my_list = []
for x in df['City']:
    # count null values
    if type(x) == float:
        city_dict['nan'] = city_dict.get('nan', 0) + 1 
        my_list.append('nan')
    else:
        len_of_list = len(x)
        city_dict[str(len_of_list)] = city_dict.get(str(len_of_list), 0) + 1 
        my_list.append(len_of_list)

In [27]:
city_dict

{'1': 260, 'nan': 236, '2': 117, '3': 29, '4': 7, '5': 1}

I now have 494 rows that either have 1 city or no cities, I previously only had 299 rows. Now lets check the instance of a length of 5

In [28]:
df.loc[(df['city_len'] == 5) & (df['City'].isnull() == False)]

Headline Country  \
428  Lower Hospitalization in Pekanbaru after Rotav...     NaN   

                                  City city_len  
428  [Man, Rota, Ho, Pekanbaru, Pekan]        5

In [29]:
df.loc[428, 'Headline']

'Lower Hospitalization in Pekanbaru after Rotavirus Vaccine becomes Mandatory'

In [30]:
df.loc[16]

Headline    New Los Angeles Hairstyle goes Viral
Country                                      NaN
City                               [Los Angeles]
city_len                                       1
Name: 16, dtype: object

In [31]:
df.head()

Headline   Country                  City  \
0                 Zika Outbreak Hits Miami       NaN               [Miami]   
1           Could Zika Reach New York City       NaN                   NaN   
2        First Case of Zika in Miami Beach       NaN  [Miami, Miami Beach]   
3   Mystery Virus Spreads in Recife Brazil  [Brazil]         [Recife, Bra]   
4  Dallas man comes down with case of Zika       NaN                   NaN   

  city_len  
0        1  
1        3  
2        2  
3        2  
4        2

In [32]:
def clean_city(headline, city):
    cities_test = []
    if (type(city) == list):
        if (len(city) > 1):
            for x in headline.split(' '):
                if x in city:
                    cities_test.append(x)
            # check if any lists have cities with more than 1 word ie Los Angeles
            max_len = max([len(x.split()) for x in city])
            for counter, x in enumerate(headline):
                for j in range(1, max_len+1):
                    my_string = ''
                    for i in range(j):
                        if counter + i < len(headline.split()):
                            my_string = my_string + f' {headline.split()[counter + i]}'
                            my_string = my_string.strip()
                            if my_string in city:
                                cities_test.append(my_string)
        else:
            return city
    res = []
    [res.append(x) for x in cities_test if x not in res]
    return res 

In [33]:
new_cities = []
for index, row in df.iterrows():
    new_cities.append(clean_city(df.loc[index, 'Headline'], df.loc[index, 'City']))

In [34]:
len(new_cities), df.shape

(650, (650, 4))

In [35]:
df['New_Cities'] = new_cities

In [40]:
df.head(25)

Headline   Country  \
0                            Zika Outbreak Hits Miami       NaN   
1                      Could Zika Reach New York City       NaN   
2                   First Case of Zika in Miami Beach       NaN   
3              Mystery Virus Spreads in Recife Brazil  [Brazil]   
4             Dallas man comes down with case of Zika       NaN   
5                   Trinidad confirms first Zika case       NaN   
6              Zika Concerns are Spreading in Houston       NaN   
7               Geneve Scientists Battle to Find Cure       NaN   
8               The CDC in Atlanta is Growing Worried       NaN   
9                  Zika Infested Monkeys in Sao Paulo       NaN   
10              Brownsville teen contracts Zika virus       NaN   
11  Mosquito control efforts in St Louis take new ...       NaN   
12  San Juan reports 1st US Zikarelated death amid...       NaN   
13                    Flu outbreak in Galveston Texas       NaN   
14                  Zika alert  Manila now threatened       NaN   
15                     Zika afflicts 7 in Iloilo City       NaN   
16               New Los Angeles Hairstyle goes Viral       NaN   
17                      Louisiana Zika cases up to 26       NaN   
18               Orlando volunteers aid Zika research       NaN   
19                Zika infects pregnant woman in Cebu       NaN   
20                  Chicago First Zika Case Confirmed       NaN   
21              Tampa Bay Area Zika Case Count Climbs       NaN   
22      Bad Water Leads to Sickness in Flint Michigan       NaN   
23                     Baltimore plans for Zika virus       NaN   
24          London Health Unit Tracks Mad Cow Disease       NaN   

                    City city_len            New_Cities  
0                [Miami]        1               [Miami]  
1                    NaN        3                    []  
2   [Miami, Miami Beach]        2  [Miami, Miami Beach]  
3          [Recife, Bra]        2              [Recife]  
4                    NaN        2                    []  
5                    NaN        3                    []  
6          [Ho, Houston]        2             [Houston]  
7                    NaN      nan                    []  
8              [Atlanta]        1             [Atlanta]  
9             [Pau, Mon]        2                    []  
10                   NaN        3                    []  
11                   NaN      nan                    []  
12                   NaN        6                    []  
13           [Galveston]        1           [Galveston]  
14         [Man, Manila]        2              [Manila]  
15         [Ilo, Iloilo]        2              [Iloilo]  
16         [Los Angeles]        1         [Los Angeles]  
17                   NaN      nan                    []  
18             [Orlando]        1             [Orlando]  
19                   NaN      nan                    []  
20             [Chicago]        1             [Chicago]  
21          [Bay, Tampa]        2          [Tampa, Bay]  
22                   NaN        3                    []  
23           [Baltimore]        1           [Baltimore]  
24                   NaN        3                    []

In [47]:
def final_city(list_cities):
    max_len = 0
    city = ''
    for x in list_cities:
        if len(x) > max_len:
            city = x
            max_len = len(x)
    return city

In [48]:
final_city(df.loc[2, 'New_Cities'])

'Miami Beach'

In [49]:
df['new_single_city'] = df['New_Cities'].map(final_city) 

In [54]:
def final_country(country):
    if type(country) == float:
        return ''
    else:
        return country[0]

In [55]:
df['new_Country'] = df['Country'].map(final_country) 

In [56]:
df.head()

Headline   Country                  City  \
0                 Zika Outbreak Hits Miami       NaN               [Miami]   
1           Could Zika Reach New York City       NaN                   NaN   
2        First Case of Zika in Miami Beach       NaN  [Miami, Miami Beach]   
3   Mystery Virus Spreads in Recife Brazil  [Brazil]         [Recife, Bra]   
4  Dallas man comes down with case of Zika       NaN                   NaN   

  city_len            New_Cities new_single_city new_Country  
0        1               [Miami]           Miami              
1        3                    []   New York City              
2        2  [Miami, Miami Beach]     Miami Beach              
3        2              [Recife]          Recife      Brazil  
4        2                    []

In [57]:
df_final = df[['Headline', 'new_Country', 'new_single_city']].copy()

In [58]:
df_final.columns = ['Headline', 'Country', 'City']

In [59]:
df_final.to_csv('part_1.csv', index = False)